### ISSUE TITLE
Hips Orthogonal basis and arm flip

### ISSUE THREAD

**Issue by Eagleton** [TemugeB/joint_angles_calculate#10]
Hi @TemugeB! Read your post and it has helped me a ton. When I wan't to use this way to other works, like MotionBert, it comes to 2 problem:
1. When you calculate the hips rotation, you select u = hips-lefthip, v = hips-neck, w = cross(u, v), and then decompose rotation matrix. But in many case, u and v are not orthogonal. What should I do in that case? Reserve u or v and use Gram-Schmidt Orthogonalization?
![image](https://github.com/TemugeB/joint_angles_calculate/assets/37317459/daac7847-f6ff-4785-a092-125cb4960c90)
2. I'm having trouble when I try to retarget to a standard skeleton in maya
My skeleton is here, it's shoulder is connected to spine, and neck is connected to spine
![image](https://github.com/TemugeB/joint_angles_calculate/assets/37317459/302e3ca8-10bf-4a38-a938-4d80b446f844)
Follow to you, I first calculate the local matrix, then calculate the world matrix by get_rotation_chain function, and then turn to euler in world space, finally, I assign the world euler to maya joints.
![image](https://github.com/TemugeB/joint_angles_calculate/assets/37317459/0f0067da-c9bf-4adf-8bc7-6945a0802e96)
The problem is the arm will flip when I retarget to the lafan1 model
![image](https://github.com/TemugeB/joint_angles_calculate/assets/37317459/c1264140-f9f3-4762-93c9-891bae96830f)
Is there any problem when convert euler or calculate joint rotation matrix? Maybe the sqrt in Decompose Function?

**Comment by TemugeB** [comment]
Hi,

Thanks for reading my post. The skeleton in my example code is defined in that way because of the keypoints that Mediapipe detects. But the ideas from my post should still carry over to other skeleton definitions.

1. The purpose of w = cross(u ,v) is to define a forward direction. In your case, u and v are not orthogonal. But w = cross(u,v) will still define a forward direction, but not a unit vector. So just convert it to a unit vector. 
Assuming v = pelvis_to_spine and w = forward_unit_vector, then recalculate u. So u = cross(v, w). This will create the unit vectors that you need to make the first rotation matrix. 

2. I think the skeleton you showed is already taking it into account, but just in case, you are changing the offsets of each joint position right? For example, some of these offsets in this function will not be single direction vectors:   
https://github.com/TemugeB/joint_angles_calculate/blob/caec22c4ede33da6b8d037e3f62a8085e6a8dc20/calculate_joint_angles.py#L128

I would suggest that you feed in all 0s angles to reproduce the Tpose first and then feed single angles to specific joints to make sure they are rotated as you expect. This usually helps me debug. 

If things still don't work, please leave another comment. I might be overlooking something and need to think more about it before I comment.

**Comment by TemugeB** [comment]
By the way, are the hip joints rigidly attached to the pelvis? If not, then the formulation above will not work. Take a look at this drawing. In the first figure, how would you define the forward direction? In the second figure, the pelvis and hips are rigidly attached so you can define a forward direction. 

In the first figure, it is still possible to define the forward direction if you take the vector between the hips as **u**, not vector from pelvis to one hip. Keep this in mind. 

![image](https://github.com/TemugeB/joint_angles_calculate/assets/36071915/7eb722bd-dd23-4a76-b894-5bba47343cdd)

**Comment by Eagleton** [comment]
I just rename the hips to pelvis, it's the same joint

**Comment by TemugeB** [comment]
Yeah I understand the joint names. If you see the left figure above, you can see that you might get very strange forward directions if you try to use pelvis_to_hips as **u**. So my suggestion is to use **u** = right_hip - left_hip.

**Comment by Eagleton** [comment]
Actually, I already did that
![image](https://github.com/TemugeB/joint_angles_calculate/assets/37317459/77a732c8-cae9-477a-a4ff-e6917d97f789)
where RHip index = 1 and LHip index = 4

**Comment by TemugeB** [comment]
frame_pos[7] is spine? 

In any case, this looks correct. If you pass in all 0 angles, do you get the correct T pose?

**Comment by Eagleton** [comment]
spine index = 7
From the point of view of the skeleton，It seems to a correct T other except pelvis.
![image](https://github.com/TemugeB/joint_angles_calculate/assets/37317459/a9779398-1e90-4479-a5cf-65a440a17c16)
But some frame, the spine seems not very good
![image](https://github.com/TemugeB/joint_angles_calculate/assets/37317459/712577e9-7aab-4609-858a-6c59233fe13a)

**Comment by TemugeB** [comment]
Looks like the hips are flat, unlike the 2nd image in your original post. Could you try to find out why the hips are not as expected?
Did you correctly set the offset for the hips?

**Comment by Eagleton** [comment]
When I calculate, I use the offset of my standard joints in the 2nd image in my original post. In my original post, I set hips world position and joints local euler, so if I pass in all 0 angles, I will always get correct T pose.  So I just apply the rotation result and all the world pos to the origin bone from motionbert result to debug T pose, maybe the debug is not correct

**Comment by TemugeB** [comment]
I'm not sure what you mean but these vectors are not in your T pose reconstruction. Can you check?
![image](https://github.com/TemugeB/joint_angles_calculate/assets/36071915/675cca26-bdb1-4233-a3d2-de45cae6344e)

**Comment by Eagleton** [comment]
I mean I have two skeleton.  One is the defined skeleton(left), and the other it the debug skeleton now
![image](https://github.com/TemugeB/joint_angles_calculate/assets/37317459/7405c94f-709e-4cd3-8447-ee0ac021aa4f)

**Comment by TemugeB** [comment]
Well, try to find out why this is happening. After that, set joint angles one by one to make sure they are rotating as you expect.

**Comment by Eagleton** [comment]
It seems that in current frame the x is not correct
the result
![image](https://github.com/TemugeB/joint_angles_calculate/assets/37317459/0d551d7b-36f3-4c0a-9a00-e78180bc0973)
modified x
![image](https://github.com/TemugeB/joint_angles_calculate/assets/37317459/0023e3ac-8ab8-4d29-a52e-d1e60aaca3dd)
